src/kakarot/accounts/library.cairo

```diff
@@ -160,38 +160,49 @@ namespace AccountContract {
 
     // EOA functions
 
-    // @notice Validate the signature of every call in the call array.
-    // @dev Recursively validates if tx is signed and valid for each call -> see utils/eth_transaction.cairo
+    // @notice Validate the signature of the call in the call array.
+    // @dev Validates if tx is signed and valid for each call -> see utils/eth_transaction.cairo
     // @param call_array_len The length of the call array.
     // @param call_array The call array.
     // @param calldata_len The length of the calldata.
     // @param calldata The calldata.
+    // @param signature_len The length of tx signature
+    // @param signature The tx signature.
+    // @param nonce The nonce of the signature
+    // @param chain_id The chain_id
     func validate{
         syscall_ptr: felt*,
         pedersen_ptr: HashBuiltin*,
         bitwise_ptr: BitwiseBuiltin*,
         range_check_ptr,
-    }(call_array_len: felt, call_array: CallArray*, calldata_len: felt, calldata: felt*) -> () {
+    }(
+        call_array_len: felt,
+        call_array: CallArray*,
+        calldata_len: felt,
+        calldata: felt*,
+        signature_len: felt,
+        signature: felt*,
+        nonce: felt,
+        chain_id: felt,
+    ) -> () {
         alloc_locals;
-        if (call_array_len == 0) {
-            // Validates that this account doesn't have code. Check only once at the end of the recursion.
-            let (bytecode_len) = Account_bytecode_len.read();
-            with_attr error_message("EOAs cannot have code") {
-                assert bytecode_len = 0;
-            }
 
-            return ();
+        // Validates that this account doesn't have code.
+        let (bytecode_len) = Account_bytecode_len.read();
+        with_attr error_message("EOAs cannot have code") {
+            assert bytecode_len = 0;
         }
 
+        with_attr error_message("Incorrect signature length") {
```

Comment:
move closed to the place where the signature is used

Reply:
done with f37cf35